### PR TITLE
Do not rely on secretRef for ManagedSeed

### LINKED ISSUE #1308: Webterminals: Shoot-terminal wrong attach info
Hi, 

we want to enable the webterminals feature in our Gardener setup. So far we were able to get the terminal feature for the (virtual) garden cluster running. For this, we needed to set the `gardenTerminalHost.apiServerIngressHost` to our api-server ingress host, since our dashboard wanted us to connect to null.ingress..., otherwise.
Now we are trying to enable the shoot terminals, but run into the issue that the websocket our browser want to connect to (k-some-hast.ingress....) is different from the api-server-url of the shoot (api.shoot.namespace...). Of course, we cannot simple specify an api-server ingress host in this case, but we are stuck with this issue. Do you have some idea what were are missing. 

Currently our dashboard (deployed via helm) comes with these values:
```yaml
ingress:                                                          
  annotations:                                                    
    cert-manager.io/cluster-issuer: letsencrypt-staging           
frontendConfig:                                                   
  seedCandidateDeterminationStrategy: MinimalDistance             
  features:                                                       
    terminalEnabled: true                                         
terminal:                                                         
  container:                                                      
    image: eu.gcr.io/gardener-project/gardener/ops-toolbelt:latest
  gardenTerminalHost:                                             
    apiServerIngressHost: OUR-HOST
    secretRef:                                                    
      namespace: garden                                           
      labelSelector:                                              
      - runtime=gardenTerminalHost                                
  garden:                                                         
    operatorCredentials:                                          
      secretRef:                                                  
        name: kubeconfig                                          
        namespace: terminal-system                                
                                                                  
  bootstrap:                                                      
    disabled: false                                               
    shootDisabled: false                                          
    seedDisabled: false                                           
    gardenTerminalHostDisabled: true                              
    apiServerIngress:                                             
      annotations:                                                
        cert-manager.io/cluster-issuer:  letsencrypt-staging                         
        cert.gardener.cloud/purpose: managed                      
        kubernetes.io/ingress.class: nginx                        
        nginx.ingress.kubernetes.io/backend-protocol: HTTPS       

```

Our shoot-api-servers are exposed as `gateway` resources and not as `ingress`es (we are running Gardener v1.56 with ManagedIstio and ApiServerSNI enabled, which became the default recently to the best of our knowledge).

### PR DESCRIPTION
**Pull request by petersutter**
**What this PR does / why we need it**:
`spec.secretRef` is not required anymore if `Seed` is a `ManagedSeed`

**Which issue(s) this PR fixes**:
Fixes #1308

**Special notes for your reviewer**:

**Release note**:

```other operator
Terminal: `spec.secretRef` is not required anymore if `Seed` is a `ManagedSeed`
```

### CODE REVIEW

File: backend/lib/services/terminals/terminalBootstrap.js [hunk 1]
```diff
@@ -17,6 +17,9 @@ const { NotImplemented } = require('http-errors')
 
 const { isHttpError } = require('@gardener-dashboard/request')
 const { dashboardClient } = require('@gardener-dashboard/kube-client')
+const {
+  getManagedSeedOrUndefined
```
Comment (holgerkoser) [inline backend/lib/services/terminals/terminalBootstrap.js:21]: I think it is a normal behavior that a function does not return something or returns null. We should not put input parameter names or possible result values into the name of the function. The name of a function should only describe what the function does. If it does not find a managed seed it cannot return one.

File: backend/lib/services/terminals/terminalBootstrap.js [hunk 2]
```diff
@@ -308,9 +310,10 @@ async function handleSeed (seed) {
 
   logger.debug(`replacing resources on seed ${name} for webterminals`)
 
+  const managedSeed = await getManagedSeedOrUndefined(dashboardClient, name)
   // now make sure a browser-trusted certificate is presented for the kube-apiserver
-  const shoot = await dashboardClient.getShoot({ namespace, name, throwNotFound: false })
-  if (shoot) {
+  if (managedSeed) {
+    const shoot = await dashboardClient.getShoot({ namespace: managedSeed.metadata.namespace, name: managedSeed.spec.shoot.name })
```
Comment (holgerkoser) [inline backend/lib/services/terminals/terminalBootstrap.js:316]:
```suggestion
    const shoot = await dashboardClient.getShoot({ 
      namespace: managedSeed.metadata.namespace, 
      name: managedSeed.spec.shoot.name 
    })
```

File: backend/lib/services/terminals/terminalBootstrap.js [hunk 3]
```diff
@@ -346,9 +349,21 @@ async function ensureTrustedCertForShootApiServer (client, shootResource, seedRe
...
   }
 
-  if (!seedResource.spec.secretRef) {
-    logger.info(`Bootstrapping Shoot ${namespace}/${name} aborted as 'spec.secretRef' on the seed is missing. In case a shoot is used as seed, add the flag \`with-secret-ref\` to the \`shoot.gardener.cloud/use-as-seed\` annotation`)
-    return
+  let seedClient
+
+  const managedSeed = await getManagedSeedOrUndefined(client, seedName)
+  if (managedSeed) {
+    const shootRef = {
```
Comment (holgerkoser) [inline backend/lib/services/terminals/terminalBootstrap.js:356]: This is also used above. May be extract a helper function:
```js
function getShootRef (managedSeed) {
   return {
      namespace: managedSeed.metadata.namespace,
      name: managedSeed.spec.shoot.name
    }
}
```

File: backend/lib/services/terminals/utils.js [hunk 4]
```diff
@@ -87,16 +89,15 @@ async function getGardenHostClusterKubeApiServer (client) {
...
 
-async function getKubeApiServerHostForSeedOrShootedSeed (client, seed) {
-  const name = seed.metadata.name
-  const namespace = 'garden'
-
-  const shoot = await client.getShoot({ namespace, name, throwNotFound: false })
-  if (shoot) {
+async function getKubeApiServerHostForSeedOrManagedSeed (client, seed, managedSeed) {
+  if (managedSeed) {
+    const name = managedSeed.spec.shoot.name
```
Comment (holgerkoser) [inline backend/lib/services/terminals/utils.js:94]: Same as above `getShootRef`

File: backend/lib/services/terminals/utils.js [hunk 5]
```diff
@@ -138,13 +139,25 @@ function getGardenTerminalHostClusterSecrets (client) {
   return client.core.secrets.list(namespace, query)
 }
 
+async function getManagedSeedOrUndefined (client, name) {
+  try {
+    return await client['seedmanagement.gardener.cloud'].managedseeds.get('garden', name) // Currently, managed seeds are restricted to the garden namespace
+  } catch (err) {
+    if (isHttpError(err, 404)) {
+      return undefined
```
Comment (holgerkoser) [inline backend/lib/services/terminals/utils.js:147]:
```suggestion
      return
```

File: backend/lib/services/terminals/terminalBootstrap.js (hunk 2 above)
Comment (petersutter) [reply backend/lib/services/terminals/terminalBootstrap.js:316]: ✅

File: backend/lib/services/terminals/terminalBootstrap.js (hunk 1 above)
Comment (petersutter) [reply backend/lib/services/terminals/terminalBootstrap.js:21]: ✅

File: backend/lib/services/terminals/terminalBootstrap.js (hunk 3 above)
Comment (petersutter) [reply backend/lib/services/terminals/terminalBootstrap.js:356]: ✅

File: backend/lib/services/terminals/utils.js (hunk 4 above)
Comment (petersutter) [reply backend/lib/services/terminals/utils.js:94]: ✅

File: backend/lib/services/terminals/utils.js (hunk 5 above)
Comment (petersutter) [reply backend/lib/services/terminals/utils.js:147]: ✅

File: backend/__fixtures__/secrets.js [hunk 6]
```diff
@@ -176,7 +175,7 @@ const secrets = {
         'gardener.cloud/role': 'ca-cluster'
       },
       data: {
-        'ca.crt': caCrt
+        'ca.crt': 'ca.crt'
```
Comment (holgerkoser) [inline backend/__fixtures__/secrets.js:178]: I thought secret data are always base64 encoded?

Comment (petersutter) [reply backend/__fixtures__/secrets.js:178]: @holgerkoser the `getSecret` helper function base64 encodes the data https://github.com/gardener/dashboard/blob/7fe8183cdf282415bac9556696c9933766a2bdf2/backend/__fixtures__/secrets.js#L20-L37

File: backend/lib/services/terminals/index.js [hunk 7]
```diff
@@ -282,22 +283,39 @@ async function getTargetCluster ({ user, namespace, name, target, preferredHost,
...
-      targetCluster.apiServer.caData = kubeconfigObject.currentCluster['certificate-authority-data']
+      const managedSeed = await client.getManagedSeed({ namespace: 'garden', name: seedName, throwNotFound: false })
+
+      let credentials
+      let caData
+      if (managedSeed) {
+        const shootRef = getShootRef(managedSeed)
+        credentials = { shootRef }
+
+        const caCluster = await client.core.secrets.get(managedSeed.metadata.namespace, `${managedSeed.spec.shoot.name}.ca-cluster`)
```
Comment (holgerkoser) [inline backend/lib/services/terminals/index.js:295]:
```suggestion
        const caCluster = await client.core.secrets.get(shootRef.namespace, `${shootRef.name}.ca-cluster`)
```

File: backend/lib/services/terminals/index.js [hunk 8]
```diff
@@ -282,22 +283,39 @@ async function getTargetCluster ({ user, namespace, name, target, preferredHost,
...
+        caData = caCluster.data['ca.crt']
+      } else {
+        const seed = getSeed(seedName)
+        const secretRef = _.get(seed, 'spec.secretRef')
```
Comment (holgerkoser) [inline backend/lib/services/terminals/index.js:299]:
```suggestion
        const secretRef = seed.spec?.secretRef
```

File: backend/lib/services/terminals/index.js [hunk 9]
```diff
@@ -348,14 +366,23 @@ async function getSeedHostCluster (client, { namespace, name, target, body, shoo
...
-  hostCluster.credentials = {
-    secretRef: _.get(seed, 'spec.secretRef')
+  let credentials
+  if (managedSeed) {
+    const shootRef = getShootRef(managedSeed)
+    credentials = { shootRef }
+  } else {
+    credentials = {
+      secretRef: _.get(seed, 'spec.secretRef')
+    }
```
Comment (holgerkoser) [inline backend/lib/services/terminals/index.js:380]:
```suggestion
    const secretRef = seed.spec?.secretRef
    credentials = { secretRef }
```

File: backend/lib/services/terminals/index.js (hunk 7 above)
Comment (petersutter) [reply backend/lib/services/terminals/index.js:295]: ✅

File: backend/lib/services/terminals/index.js (hunk 9 above)
Comment (petersutter) [reply backend/lib/services/terminals/index.js:380]: ✅

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Review by grolu — APPROVED** [review]
/lgtm

**Pull request merged**